src/inference/include/openvino/runtime/async_infer_queue.hpp

```diff
@@ -0,0 +1,124 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/**
+ * @brief A header file that provides wrapper classes for infer requests and callbacks.
+ *
+ * @file openvino/runtime/async_infer_queue.hpp
+ */
+#pragma once
+
+#include <condition_variable>
+#include <map>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <vector>
+
+#include "openvino/runtime/common.hpp"
+#include "openvino/runtime/executable_network.hpp"
+#include "openvino/runtime/infer_request.hpp"
+#include "openvino/runtime/tensor.hpp"
+
+namespace ov {
+namespace runtime {
+
+/**
+ * @brief This is an interface of asynchronous infer request
+ *
+ * It can throw exceptions safely for the application, where it is properly handled.
+ *
+ * @note AsyncInferQueue provides simple and unifed interface for multiple asynchronous
+ * calls to OpenVINO's InferRequests.
+ */
+class OPENVINO_RUNTIME_API AsyncInferQueue {
+public:
+    AsyncInferQueue() = delete;
+
+    /**
+     * @brief Basic constructor. Creates InferRequests based on given ExecutableNetwork.
+     *
+     * @param net ExecutableNetwork object which will be base for creating of InferRequests.
+     * @param jobs Number of InferRequests to be created. If equals to @p 0 AsyncInferQueue
+     * will automatically try to choose number of jobs.
+     */
+    AsyncInferQueue(ExecutableNetwork& net, size_t jobs = 0);
+
+    // ~AsyncInferQueue() {
+    //     _requests.clear();
+    // }
+
+    /**
+     * @brief Starts asynchronous inference on next avaiable job. Blocking call if there are no free jobs.
+     *
+     * @param userdata Data passed to InferRequest's callback. Can be anything that matches ov::Any.
+     */
+    void start_async(const ov::Any userdata = nullptr);
+
+    /**
+     * @brief Starts asynchronous inference with given input data. Blocking call if there are no free jobs.
+     *
+     * @param inputs Map containing integers (index) and Tensor pairs.
+     * @param userdata Data passed to InferRequest's callback. Can be anything that matches ov::Any.
+     */
+    void start_async(std::map<size_t, ov::runtime::Tensor> inputs, const ov::Any userdata = nullptr);
+
+    /**
+     * @brief Starts asynchronous inference with given input data. Blocking call if there are no free jobs.
+     *
+     * @param inputs Map containing strings (names) and Tensor pairs.
+     * @param userdata Data passed to InferRequest's callback. Can be anything that matches ov::Any.
+     */
+    void start_async(std::map<std::string, ov::runtime::Tensor> inputs, const ov::Any userdata = nullptr);
+
+    /**
+     * @brief Waits for any request to become avaiable. Blocking call if there are no free jobs.
+     *
+     * @note Waits for InferRequest while queue is running.
+     * @return Returns @p true if at least one request is avaiable, otherwise @p false.
+     */
+    bool is_ready();
```

Comment:
What about merge `is_ready` and `get_idle_id` and `operator[]` into one `bool try_pop(InferRequest&)`  method?

Reply:
Sorry. missed that is ready is blocking. So it should  be blocking `InferRequest& pop()` method

Reply:
Sure `is_ready()` is blocking, however `pop_idle_id()` this method is only implementation detail, it is not public interface. Maybe I should move it to protected, shouldn't I?